modules/rostests/apitests/user32/PrivateExtractIcons.c

```diff
@@ -3,38 +3,132 @@
  * LICENSE:         LGPLv2.1+ - See COPYING.LIB in the top level directory
  * PURPOSE:         Test for PrivateExtractIcons
  * PROGRAMMER:      Hermes Belusca-Maito
+ *                  Doug Lyons <douglyons@douglyons.com>
  */
 
 #include "precomp.h"
+#include <stdio.h>
+
+BOOL FileExists(LPCSTR FileName)
+{
+    DWORD Attribute = GetFileAttributes(FileName);
+
+    return (Attribute != INVALID_FILE_ATTRIBUTES &&
+            !(Attribute & FILE_ATTRIBUTE_DIRECTORY));
+}
+
+BOOL ResourceToFile(INT i, LPCSTR FileName)
+{
+    FILE *fout;
+    HGLOBAL hData;
+    HRSRC hRes;
+    LPVOID lpResLock;
+    UINT iSize;
+
+    if (FileExists(FileName))
+    {
+        skip("'%s' already exists. Exiting now\n", FileName);
+        return FALSE;
+    }
+
+    hRes = FindResourceW(NULL, MAKEINTRESOURCEW(i), MAKEINTRESOURCEW(RT_RCDATA));
+    if (hRes == NULL)
+    {
+        skip("Could not locate resource (%d). Exiting now\n", i);
+        return FALSE;
+    }
+
+    iSize = SizeofResource(NULL, hRes);
+
+    hData = LoadResource(NULL, hRes);
+    if (hData == NULL)
+    {
+        skip("Could not load resource (%d). Exiting now\n", i);
+        return FALSE;
+    }
+
+    // Lock the resource into global memory.
+    lpResLock = LockResource(hData);
+    if (lpResLock == NULL)
+    {
+        skip("Could not lock resource (%d). Exiting now\n", i);
+        return FALSE;
+    }
+
+    fout = fopen(FileName, "wb");
+    fwrite(lpResLock, iSize, 1, fout);
+    fclose(fout);
+    return TRUE;
+}
 
 static struct
 {
     PCWSTR FilePath;
-    UINT cIcons;        // Return value from PrivateExtractIconsW
+    UINT cIcons;        // Return value of the first icon group extracted (should be 1 if no error)
+    UINT cTotalIcons;   // Return value of total icon groups in file
     BOOL bhIconValid;   // Whether or not the returned icon handle is not NULL.
 } IconTests[] =
 {
-    /* Executables with icons */
-    {L"notepad.exe", 1, TRUE},
-    {L"%SystemRoot%\\System32\\cmd.exe", 1, TRUE},
+    /* Executables with just one icon group */
+    {L"notepad.exe", 1, 1, TRUE},
+    {L"%SystemRoot%\\System32\\cmd.exe", 1, 1, TRUE},
 
-    /* Executable without icon */
-    {L"%SystemRoot%\\System32\\autochk.exe", 0, FALSE},
+    /* Executable without icon groups */
+    {L"%SystemRoot%\\System32\\autochk.exe", 0, 0, FALSE},
 
-    /* Existing file */
-    {L"%SystemRoot%\\System32\\shell32.dll", 1, TRUE},
+    /* Existing file (shell32 has 233 icon groups in ReactOS only) */
```

Comment:
I did a quick search for something that might be a good fit for what you asked, but nothing was obvious. This would be a 'nice to have' feature though if someone finds a good candidate. Maybe for another PR? Thanks.

Reply:
no problem :+1: